src/coreclr/jit/fgopt.cpp

```diff
@@ -2389,6 +2389,15 @@ void Compiler::fgCompactBlocks(BasicBlock* block, BasicBlock* bNext)
             break;
     }
 
+    if ((bNext->bbJumpDest != nullptr) && bNext->bbJumpDest->isLoopAlign())
+    {
+        // `bNext` has a backward target to some block which mean bNext is part of a loop.
+        // `block` into which `bNext` is compacted should be updated with its loop number
+        JITDUMP("Updating loop number for " FMT_BB " from " FMT_LP " to " FMT_LP ".\n", block->bbNum,
+                block->bbNatLoopNum, bNext->bbNatLoopNum);
+        block->bbNatLoopNum = bNext->bbNatLoopNum;
+    }
+
```

Comment:
Should we also update the loop table -- presumably this loop now has a new top? Or can this only happen when we no longer are maintaining the loop table?

Reply:
We do update loop table by calling `fgUpdateLoopsAfterCompacting()` below.